dagshub/auth/oauth.py

```diff
@@ -0,0 +1,126 @@
+from typing import Optional, Dict
+import appdirs
+import datetime
+import getpass
+import logging
+import os
+import requests
+import traceback
+import urllib
+import uuid
+import yaml
+
+HOST_KEY = "DAGSHUB_CLIENT_HOST"
+DEFAULT_HOST = "https://dagshub.com"
+CACHE_LOCATION_KEY = "DAGSHUB_CLIENT_CONFIG"
+DEFAULT_CACHE_LOCATION = os.path.join(appdirs.user_cache_dir("dagshub"), "tokens")
+CLIENT_ID_KEY = "DAGSHUB_CLIENT_ID"
+DEFAULT_CLIENT_ID = "32b60ba385aa7cecf24046d8195a71c07dd345d9657977863b52e7748e0f0f28"
+
+logger = logging.getLogger(__name__)
+
+
+class OAuthAuthenticator:
+    def __init__(self, **kwargs):
+        self.host: str = kwargs.get("host", os.environ.get(HOST_KEY, DEFAULT_HOST)).strip("/")
+        self.client_id: str = kwargs.get(
+            "client_id", os.environ.get(CLIENT_ID_KEY, DEFAULT_CLIENT_ID)
+        )
+        self.cache_location: str = kwargs.get(
+            "cache_location", os.environ.get(CACHE_LOCATION_KEY, DEFAULT_CACHE_LOCATION)
+        )
+        self._token_cache: Optional[Dict[str, Dict]] = None
+
+    def get_oauth_token(self) -> str:
+        if self._token_cache is None:
+            self._token_cache = self._load_cache_file()
+        token = self._token_cache.get(self.host, None)
+        if token is None or self._is_expired(token):
+            token = self.oauth_flow()
+            self._token_cache[self.host] = token
+            self._store_cache_file()
+        return token["access_token"]
+
+    def oauth_flow(self) -> Dict:
+        state = uuid.uuid4()
+        dagshub_url = urllib.parse.urljoin(self.host, "login/oauth")
+        print(
+            f"Go to {dagshub_url}/authorize?state={state}&client_id={self.client_id} and paste the code back in here."
+        )
+        code = getpass.getpass("Code:")
```

Comment:
Since this code SHOULD get executed in CI\CD scenarios, is there a standard way for you to check whether you're in an interactive session, and if not then just crash instead of getting the machine stuck on an input and wasting user cloud credits?

Reply:
Will research more